### PR TITLE
perf_nmem.py: Add nmem PMU test cases

### PR DESCRIPTION
**Pull request by nasastry**
Add nmem PMU test cases

Check for nmem PMUs in sysfs
Check for CONFIG_PAPR_SCM kernel configuration
Check for nmem PMU register messages in dmesg
Check PMUs found in sysfs and dmesg are matching
Count the number of PMU events
Run all events one by one per PMU
Run events by grouping per PMU
Run mixed PMU events - negative test case
Check PMU cpumask exists or not
Check cpumask is same for all the PMUs or not
Check cpumask by on and off cpu

Signed-off-by: Nageswara R Sastry <rnsastry@linux.ibm.com>

### CODE REVIEW

**Comment by nasastry** [comment]
```
JOB ID     : 282eba676ae371a68efd6c3414c130df418fc8a8
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-06-25T07.34-282eba6/job.log
 (01/10) perf_nmem.py:perfNMEM.test_kernel_config: PASS (0.60 s)
 (02/10) perf_nmem.py:perfNMEM.test_pmu_register_dmesg: PASS (0.68 s)
 (03/10) perf_nmem.py:perfNMEM.test_sysfs: PASS (0.83 s)
 (04/10) perf_nmem.py:perfNMEM.test_pmu_count: PASS (0.60 s)
 (05/10) perf_nmem.py:perfNMEM.test_all_events: PASS (42.33 s)
 (06/10) perf_nmem.py:perfNMEM.test_all_group_events: PASS (3.06 s)
 (07/10) perf_nmem.py:perfNMEM.test_mixed_events: PASS (1.83 s)
 (08/10) perf_nmem.py:perfNMEM.test_pmu_cpumask: PASS (0.60 s)
 (09/10) perf_nmem.py:perfNMEM.test_cpumask: PASS (0.60 s)
 (10/10) perf_nmem.py:perfNMEM.test_cpumask_cpu_off: PASS (1.11 s)
RESULTS    : PASS 10 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-06-25T07.34-282eba6/results.html
JOB TIME   : 53.00 s
```

```
JOB ID     : e88c909006f40a397770384c982da562a03e3687
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-06-25T07.30-e88c909/job.log
 (01/10) perf_nmem.py:perfNMEM.test_kernel_config: CANCEL: nmem PMUs not found (2.00 s)
 (02/10) perf_nmem.py:perfNMEM.test_pmu_register_dmesg: CANCEL: nmem PMUs not found (1.95 s)
 (03/10) perf_nmem.py:perfNMEM.test_sysfs: CANCEL: nmem PMUs not found (2.05 s)
 (04/10) perf_nmem.py:perfNMEM.test_pmu_count: CANCEL: nmem PMUs not found (1.98 s)
 (05/10) perf_nmem.py:perfNMEM.test_all_events: CANCEL: nmem PMUs not found (2.00 s)
 (06/10) perf_nmem.py:perfNMEM.test_all_group_events: CANCEL: nmem PMUs not found (1.98 s)
 (07/10) perf_nmem.py:perfNMEM.test_mixed_events: CANCEL: nmem PMUs not found (2.01 s)
 (08/10) perf_nmem.py:perfNMEM.test_pmu_cpumask: CANCEL: nmem PMUs not found (2.09 s)
 (09/10) perf_nmem.py:perfNMEM.test_cpumask: CANCEL: nmem PMUs not found (2.01 s)
 (10/10) perf_nmem.py:perfNMEM.test_cpumask_cpu_off: CANCEL: nmem PMUs not found (2.04 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 10
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-06-25T07.30-e88c909/results.html
JOB TIME   : 21.83 s
```

**Comment by nasastry** [comment]
After removing unused variable.

```
JOB ID     : 898aa8246a9b124b3258db436f5206e2611e9c3d
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-06-26T10.24-898aa82/job.log
 (01/10) perf_nmem.py:perfNMEM.test_kernel_config: PASS (0.61 s)
 (02/10) perf_nmem.py:perfNMEM.test_pmu_register_dmesg: PASS (0.68 s)
 (03/10) perf_nmem.py:perfNMEM.test_sysfs: PASS (0.60 s)
 (04/10) perf_nmem.py:perfNMEM.test_pmu_count: PASS (0.60 s)
 (05/10) perf_nmem.py:perfNMEM.test_all_events: PASS (42.30 s)
 (06/10) perf_nmem.py:perfNMEM.test_all_group_events: PASS (3.06 s)
 (07/10) perf_nmem.py:perfNMEM.test_mixed_events: PASS (1.83 s)
 (08/10) perf_nmem.py:perfNMEM.test_pmu_cpumask: PASS (0.61 s)
 (09/10) perf_nmem.py:perfNMEM.test_cpumask: PASS (0.61 s)
 (10/10) perf_nmem.py:perfNMEM.test_cpumask_cpu_off: PASS (1.02 s)
RESULTS    : PASS 10 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-06-26T10.24-898aa82/results.html
JOB TIME   : 52.63 s
```


```
JOB ID     : 66886b08ee3fa2fc5acf7376ae8eba3132f03cfe
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-06-26T10.22-66886b0/job.log
 (01/10) perf_nmem.py:perfNMEM.test_kernel_config: CANCEL: nmem PMUs not found (2.05 s)
 (02/10) perf_nmem.py:perfNMEM.test_pmu_register_dmesg: CANCEL: nmem PMUs not found (1.98 s)
 (03/10) perf_nmem.py:perfNMEM.test_sysfs: CANCEL: nmem PMUs not found (2.02 s)
 (04/10) perf_nmem.py:perfNMEM.test_pmu_count: CANCEL: nmem PMUs not found (2.02 s)
 (05/10) perf_nmem.py:perfNMEM.test_all_events: CANCEL: nmem PMUs not found (1.99 s)
 (06/10) perf_nmem.py:perfNMEM.test_all_group_events: CANCEL: nmem PMUs not found (1.96 s)
 (07/10) perf_nmem.py:perfNMEM.test_mixed_events: CANCEL: nmem PMUs not found (2.02 s)
 (08/10) perf_nmem.py:perfNMEM.test_pmu_cpumask: CANCEL: nmem PMUs not found (1.99 s)
 (09/10) perf_nmem.py:perfNMEM.test_cpumask: CANCEL: nmem PMUs not found (2.02 s)
 (10/10) perf_nmem.py:perfNMEM.test_cpumask_cpu_off: CANCEL: nmem PMUs not found (2.01 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 10
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-06-26T10.22-66886b0/results.html
JOB TIME   : 21.76 s
```

File: perf/perf_nmem.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        # Checking all the PMUs cpumask is same or not
+        self._check_cpumask()
+
+    def _cpu_on_off(self, cpu_number, disable_flag=True):
+        cpu_file = "/sys/bus/cpu/devices/cpu%s/online" % cpu_number
+        if disable_flag:
+            genio.write_one_line(cpu_file, "0")
+            self.log.info("Offline CPU: %s" % cpu_number)
+        else:
+            genio.write_one_line(cpu_file, "1")
```
Comment (PraveenPenguin) [inline perf/perf_nmem.py:232]: using avocado CPU module will be good here

File: perf/perf_nmem.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+            self.log.info("Online CPU: %s" % cpu_number)
```
Comment (PraveenPenguin) [inline perf/perf_nmem.py:233]: using avocado CPU module will be good here as we already using CPU module

File: perf/perf_nmem.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                         % platform.uname()[2]])
+        elif detected_distro.name in ['rhel', 'SuSE', 'fedora', 'centos']:
+            deps.extend(['perf'])
+        else:
+            self.cancel("Install the package for perf supported by %s"
+                        % detected_distro.name)
+        for package in deps:
+            if not smm.check_installed(package) and not smm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+
```
Comment (PraveenPenguin) [inline perf/perf_nmem.py:60]: can this white space removed

File: perf/perf_nmem.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        # Set required variables
+        self.base_dir = "/sys/devices/"
+        self.pmu_list = []
+        self.all_events = {}
+
```
Comment (PraveenPenguin) [inline perf/perf_nmem.py:65]: here as well white space please

File: perf/perf_nmem.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+                for pmu in self.pmu_list:
+                    if ln.split('/')[0] in pmu:
+                        # Collecting the events per PMU
+                        # ex:{nmem0: [ev1, ev2, ...], nmem1: [ev1, ev2, ...]}
+                        self.all_events.setdefault(pmu, []).append(ln)
+        if not self.all_events:
+            self.cancel("Can not extract events from 'perf list nmem'")
+
+    def test_kernel_config(self):
+        cfg_param = "CONFIG_PAPR_SCM"
```
Comment (PraveenPenguin) [inline perf/perf_nmem.py:95]: one help is this is prerequisite of test or just a one test scenario ?

**Review by PraveenPenguin — CHANGES_REQUESTED** [review]

**Comment by nasastry** [comment]
@PraveenPenguin resolved the changes requested. After pushing the new code here is the run result.

```
JOB ID     : 15ac20510171f992e3949345d339772a03acc96a
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-06-29T04.54-15ac205/job.log
 (1/9) perf_nmem.py:perfNMEM.test_pmu_register_dmesg: FAIL: NVDIMM PMUs not found in dmesg. (0.65 s)
 (2/9) perf_nmem.py:perfNMEM.test_sysfs: PASS (0.60 s)
 (3/9) perf_nmem.py:perfNMEM.test_pmu_count: PASS (0.60 s)
 (4/9) perf_nmem.py:perfNMEM.test_all_events: PASS (42.32 s)
 (5/9) perf_nmem.py:perfNMEM.test_all_group_events: PASS (3.06 s)
 (6/9) perf_nmem.py:perfNMEM.test_mixed_events: PASS (1.83 s)
 (7/9) perf_nmem.py:perfNMEM.test_pmu_cpumask: PASS (0.60 s)
 (8/9) perf_nmem.py:perfNMEM.test_cpumask: PASS (0.60 s)
 (9/9) perf_nmem.py:perfNMEM.test_cpumask_cpu_off: PASS (0.99 s)
RESULTS    : PASS 8 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-06-29T04.54-15ac205/results.html
JOB TIME   : 51.88 s
```

```
JOB ID     : 7d1b95d039879683c628d0dfe785f3a7922b75be
JOB LOG    : /home/avocado-fvt-wrapper/results/job-2021-06-29T04.43-7d1b95d/job.log
 (1/9) perf_nmem.py:perfNMEM.test_pmu_register_dmesg: CANCEL: nmem PMUs not found (0.94 s)
 (2/9) perf_nmem.py:perfNMEM.test_sysfs: CANCEL: nmem PMUs not found (0.97 s)
 (3/9) perf_nmem.py:perfNMEM.test_pmu_count: CANCEL: nmem PMUs not found (0.93 s)
 (4/9) perf_nmem.py:perfNMEM.test_all_events: CANCEL: nmem PMUs not found (1.04 s)
 (5/9) perf_nmem.py:perfNMEM.test_all_group_events: CANCEL: nmem PMUs not found (0.89 s)
 (6/9) perf_nmem.py:perfNMEM.test_mixed_events: CANCEL: nmem PMUs not found (0.94 s)
 (7/9) perf_nmem.py:perfNMEM.test_pmu_cpumask: CANCEL: nmem PMUs not found (0.93 s)
 (8/9) perf_nmem.py:perfNMEM.test_cpumask: CANCEL: nmem PMUs not found (0.94 s)
 (9/9) perf_nmem.py:perfNMEM.test_cpumask_cpu_off: CANCEL: nmem PMUs not found (0.93 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 9
JOB HTML   : /home/avocado-fvt-wrapper/results/job-2021-06-29T04.43-7d1b95d/results.html
JOB TIME   : 9.19 s
```

File: perf/perf_nmem.py (hunk 5 above)
Comment (nasastry) [reply perf/perf_nmem.py]: Should be pre-requisite. changed the code.

File: perf/perf_nmem.py (hunk 4 above)
Comment (nasastry) [reply perf/perf_nmem.py]: done

File: perf/perf_nmem.py (hunk 3 above)
Comment (nasastry) [reply perf/perf_nmem.py]: done

File: perf/perf_nmem.py (hunk 2 above)
Comment (nasastry) [reply perf/perf_nmem.py]: done

**Review by PraveenPenguin — APPROVED** [review]

**Pull request merged**